### ISSUE TITLE
psycopg_escape_string loses sync with standard_conforming_strings

### ISSUE THREAD

**Issue by jcflack** [psycopg/psycopg2#366]
A colleague drew my attention to this.

`psycopg_escape_string` wants to produce the correct form for 3 cases:
- old server, no `scs`: `'`_content-with-bash-escapes_`'` (call this Form 1)
- newer server, `scs` off: `E'`_content-with-bash-escapes_`'` (Form 2)
- newer server, `scs` on: `'`_content-without-bash-escapes_`'` (Form 3)

There is really some flexibility here: in the first case, only Form 1 is valid. In the second case, Form 1 or Form 2 (with Form 1 being deprecated), and in the third, Form 2 or Form 3. But anyway....

It uses `PQescapeString...` to handle the content, automatically producing with-or-without-bash-escapes depending on the server setting of `standard_conforming_strings` (`libpq` always knows the current setting, because the server sends a message whenever it changes, and `libpq` tracks it).

`psycopg_escape_string` uses its own logic for whether to put the `E` in front, based on checking the `standard_conforming_strings` setting _once_ at connection init.

The result is, if `scs` is off when the connection is made, but a later SQL command turns it on, psycopg will keep on producing `E'` strings, but _without_ bash-escaping of the contents, which is wrong. Or, if `scs` is on at connection time, but a later SQL command turns it off, psycopg will keep producing non-`E'` strings, but _with_ bash-escaped contents ... IOW, Form 1, which is not flat-out wrong, but deprecated and invites a warning.

In my colleague's case, he didn't have control over the extra SQL command changing the `standard_conforming_strings` setting, as it was issued by another library between his code and psycopg. But clearly, psycopg can't go letting its quote style get out of sync with how the server is set.

I can think of two fix approaches:
1. Continue using `PQescapeString...` for the escaping, and the setting of `standard_conforming_strings` for deciding whether to `E` or not, but check `standard_conforming_strings` every time. This sounds expensive but it's really not, because of the way `libpq` always knows the current setting on the server, so `PQparameterStatus` is only returning a local value, not interacting with the server.
2. Ditch `PQescapeString...` and its changeable behavior, bring back the `PSYCOPG_OWN_QUOTING` code, and escape all string content the same way all the time. Emit it always with `E` on any server new enough to understand it, and without `E` only on old servers that don't. (In other words, keep the one check for `standard_conforming_strings` at connection time, but only to remember whether the parameter exists or not. If it exists, always use `E` no matter how it is set; if does not exist, don't use `E`.)

The second choice might be the simplest, KISS, no-moving-parts, nothing-to-break approach, but either way would work.

**Comment by dvarrazzo** [comment]
2 Is not going to happen.

1 is a maybe, because I don't want to query the libpq every time a string is to be escaped. I would probably be fine to query `PQparameterStatus` at every execute/executemany/callproc... but just.

We could try and fix the problem for next bugfix release. If you proposed a patch it would be better though (with tests please). Otherwise, to let you know, this issue has low priority as this code has been in production for some 10 years and this is the first time this issue crops up.

**Comment by jcflack** [comment]
Hmm, well, if the concern with 1 is efficiency, the most direct approach is to allow including `libpq-int.h` in one source file, probably `utils.c`, so it can just refer to the boolean `pgconn->std_strings` directly. No calling `PQparameterStatus`, no walking linked lists or comparing string keys, just dereferencing to one boolean, and `libpq` does the work of making sure it always reflects the true state of the connection. Doesn't get any simpler than that.

Naturally `libpq-int.h` warns that code including it becomes more sensitive to changes between PostgreSQL versions, but by being disciplined about it, only including it in one source file where it's needed, and only referring to `std_strings` in `struct pg_conn`, that concern is minimal as a practical matter. Not likely that `std_strings` will be going away any foreseeable time, and it has already been there since 8.2, 8.1.4, 8.0.8, and 7.4.13.

There is one thing I'd need to learn about `psycopg` coding practices before I could make a good patch: in your docs I read that threading is supposed to be well supported. Of course between the call to `PQescapeString...` and the check of `pgconn->std_strings` to determine what quotes to put around it, other uses of the connection have to be excluded ... what are the mechanisms used in `psycopg` for doing that?

**Comment by dvarrazzo** [comment]
No, concern 1 is not efficiency: walking a short linked list once per execution is no problem compared to the cost of the network roundtrip, we can afford that no problem instead of accessing implementation details of the libpq. So `PQparameterStatus` must be the way.

About concurrency it's no problem: only one thread at time can execute a query, there is a lock that is not released between query preparation and execution.

**Comment by dvarrazzo** [comment]
I would like to wrap the 2.6.2 release together and this change is a bit too invasive I'm afraid, so I prefer postponing it.

**Comment by dvarrazzo** [comment]
I won't fix this for psycopg 2.7; probably it will happen later. The bug hasn't been reported before in several years, so it doesn't seem affecting many people.

Note that you can start the connection with scs off passing the parameter at connection time, either in the connection string or using an env var.

```
$ python -c "import psycopg2; cnn = psycopg2.connect(''); cur = cnn.cursor(); print cur.mogrify('%s', ['hi'])"
'hi'

$ python -c "import psycopg2; cnn = psycopg2.connect(options='-c standard_conforming_strings=off'); cur = cnn.cursor(); print cur.mogrify('%s', ['hi'])"
E'hi'

$ PGOPTIONS='-c standard_conforming_strings=off' python -c "import psycopg2; cnn = psycopg2.connect(''); cur = cnn.cursor(); print cur.mogrify('%s', ['hi'])"
E'hi'
```

I think there is enough to work around the issue.
